src/zedbit/mailx/Mailx.java

```diff
@@ -0,0 +1,506 @@
+    static Set<String> urlsReachable = new TreeSet<String>(String.CASE_INSENSITIVE_ORDER); // Tracks pages where crawling is OK
+    static boolean verbose; // Are we running in verbose mode?
+    static boolean trace; // Are we running in trace mode? Useful for debugging
+
+    //Prepare to match on regexes for likely emails, very loose, to not miss domains such as .museum :)
+    static final Pattern emailP = Pattern.compile("(.*)(\\b[\\S^@]+@[\\S^@]+\\b)(.*)");
+
+    //We will follow two kinds of links, relative which we can build a regex for now, and absolute...
+    static final Pattern relHrefP = Pattern.compile("(^.*)(href=\"/)([^\"]+)(.*$)");
+    static Pattern absHrefP; // compiled at method processArgs once we have extracted our uri
+    static Pattern websiteP; // as above, used to check dynamic links/routes to be within this website
+
+    //Files we will not scan
+    static final Pattern skipP = Pattern.compile(".*(\\.(ico|jpg|jpeg|png|xml|php|php\\?rsd|css|pdf|doc|docx)|(/feed/)|(\\.css\\?.*))$", Pattern.CASE_INSENSITIVE);
+
+    // Some constants for some pretty-printing 
+    static final String ANSI_RED = "\u001B[31m";
+    static final String ANSI_RESET = "\u001B[0m";
+    static final String MARGIN = "        ";
+
+    /**
+    * Entry point for Mailx, a simple scrapping sample program for email extraction.
+    *
+    * arguments contains the commands line.
+    * Can have the following form:
+    *
+    * <p> ...Mailx <an uri to crawl from> [-v|-t], i.e. ...Mailx http://mysite.com -v </p>
+    * <p> The Optional argument -v can be provided for verbose output, -t for trace/debug output </p>
+    */
+    public static void main(String[] arguments) {
+
+        // Verify the command line arguments and load into variables startPage & uri
+        processArgs(arguments);
+
+        // Get a headless browser
+        setWebClient();
+
+        // Now begin the drill down at startPage!
+        crawl(startPage);
+
+        System.out.println ("Crawling completed!");
+
+    } // end main
+
+
+   
+    /**
+    * Crawl but limit ourselves to only links that we have not visited before
+    * <code>Crawl(String pageUrl)</code> is called 
+    * for pages for which we have a static URL. 
+    */
+    private static void crawl(String pageUrl) {
+
+
+        // First, check if we have been at this URL before, we back out
+        if (urlsVisited.contains(pageUrl) == true) {
+            printlnT("We have already visited " + pageUrl + ". Skipping it");
+            return;
+        }
+
+        // Connect and get the page, backout if we can't.
+        printlnV("Crawling Static Link w/URL=" + pageUrl);
+        HtmlPage page;  // This will be the page we will be working on, as represented by htmlUnit
+        try { page = webClient.getPage(pageUrl); }
+        catch (Exception e) {
+            // Mark we have been here before even if we got an error!
+            urlsVisited.add(pageUrl);
```

Comment:
Hi Chris, 

Yup, the real problem was not encapsulating urlsVisited and using it for two purposes, keeping urls that were visited, and those that had errors (and really do not need to be visited again). I did some rewriting that included creating a new set, urlsErrored, and encapsulated urlsVisited and urlsErrored in its own class UrlTracker --that way, this class could be expanded to include stuff such as number of emails in page, size of page, etc, and have everything be underwraps for the rest of the logic.